datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -865,9 +846,16 @@ pub async fn from_substrait_rel(
                 let name = filename.unwrap();
                 // directly use unwrap here since we could determine it is a valid one
                 let table_reference = TableReference::Bare { table: name.into() };
-                let t = ctx.table(table_reference).await?;
-                let t = t.into_optimized_plan()?;
-                extract_projection(t, &read.projection)
+                let t = ctx.table(table_reference.clone()).await?;
+
+                let substrait_schema =
+                    from_substrait_named_struct(named_struct, extensions)?
+                        .replace_qualifier(table_reference);
+
+                let substrait_schema =
```

Comment:
actually I guess all of this could be shared by all ReadRels? no need to do that refactor here if you don't want to but could be nice 🤷

Reply:
yeah, this was deliberate to keep the PR smaller. I'll do the refactor separately if you don't mind. VirtualTable is an odd case, you don't really need schema compatibility check there, so better to leave that design to a different PR imho.